libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/route/ReplayRouteSession.kt

```diff
@@ -0,0 +1,201 @@
+package com.mapbox.navigation.core.replay.route
+
+import android.annotation.SuppressLint
+import android.content.Context
+import com.mapbox.android.core.location.LocationEngineCallback
+import com.mapbox.android.core.location.LocationEngineProvider
+import com.mapbox.android.core.location.LocationEngineResult
+import com.mapbox.android.core.permissions.PermissionsManager
+import com.mapbox.api.directions.v5.DirectionsCriteria
+import com.mapbox.navigation.base.ExperimentalPreviewMapboxNavigationAPI
+import com.mapbox.navigation.base.route.NavigationRoute
+import com.mapbox.navigation.core.MapboxNavigation
+import com.mapbox.navigation.core.directions.session.RoutesObserver
+import com.mapbox.navigation.core.lifecycle.MapboxNavigationApp
+import com.mapbox.navigation.core.lifecycle.MapboxNavigationObserver
+import com.mapbox.navigation.core.replay.MapboxReplayer
+import com.mapbox.navigation.core.replay.history.ReplayEventBase
+import com.mapbox.navigation.core.replay.history.ReplayEventUpdateLocation
+import com.mapbox.navigation.core.replay.history.ReplayEventsObserver
+import com.mapbox.navigation.utils.internal.logW
+import java.util.Collections
+
+/**
+ * Used to create a replay trip session. Continue to use [MapboxNavigation.setNavigationRoutes] to
+ * decide the route that should be replayed.
+ *
+ * Do not use this class with the [ReplayProgressObserver]. They will create conflicts with and
+ * the results are undefined.
+ *
+ * The simulated driver from [ReplayRouteSession] will slow down to a stop depending
+ * [ReplayRouteSessionOptions.decodeMinDistance]. To remove this behavior you can set it to
+ * [Double.MAX_VALUE], be aware that it will require more memory.
+ *
+ * Use [ReplayRouteSessionOptions] for customizations. For example, this is how can update the
+ * location frequency.
+ *
+ * ```
+ * replayRouteSession.setOptions(
+ *     replayRouteSession.getOptions().toBuilder()
+ *         .replayRouteOptions(
+ *             replayRouteSession.getOptions().replayRouteOptions.toBuilder()
+ *                 .frequency(25.0)
+ *                 .build()
+ *         )
+ *         .build()
+ * )
+ * ```
+ *
+ * Enable and disable the [ReplayRouteSession] with [MapboxNavigation] or [MapboxNavigationApp].
+ * The replay session will be enabled when [MapboxNavigation] is attached.
+ */
+@ExperimentalPreviewMapboxNavigationAPI
+class ReplayRouteSession : MapboxNavigationObserver {
```

Comment:
We have users who still create `MapboxNavigaiton` using constructor. Will they be able to use this lazy mechanism without migrating to `MapboxNavigationApp`?

Reply:
Yeah. The `MapboxNavigationObserver` can be used without `MapboxNavigationApp`. The interface exposes the functions.

``` kotlin
interface MapboxNavigationOberver {
  fun onAttached(mapboxNavigation: MapboxNavigation)
  fun onDetached(mapboxNavigation: MapboxNavigation)
}
```

This means you can use `ReplayRouteSession` in this way. Call start and stop replay whenever you have an instance of `MapboxNavigation`. But again, this is NOT RECOMMENDED because we want to share this object with android auto.

``` kotlin
val replayRouteSession = ReplayRouteSession()

fun startReplay() {
  replayRouteSession.onAttached(mapboxNavigation)(
}

fun stopReplay() {
  replayRouteSession.onDetached(mapboxNavigation)(
}
```

Reply:
Thanks!